go/vt/wrangler/stream_migrater.go

```diff
@@ -0,0 +1,604 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package wrangler
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"golang.org/x/net/context"
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/binlog/binlogplayer"
+	"vitess.io/vitess/go/vt/concurrency"
+	"vitess.io/vitess/go/vt/key"
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/throttler"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vttablet/tabletmanager/vreplication"
+)
+
+type streamMigrater struct {
+	mi *migrater
+}
+
+type vrStream struct {
+	id       uint32
+	workflow string
+	bls      *binlogdatapb.BinlogSource
+	pos      mysql.Position
+}
+
+func (sm *streamMigrater) stopStreams(ctx context.Context) ([]*vrStream, error) {
+	if sm.mi.migrationType == binlogdatapb.MigrationType_TABLES {
+		// Source streams should be stopped only for shard migrations.
+		return nil, nil
+	}
+	streams, err := sm.readSourceStreams(ctx)
+	if err != nil {
+		return nil, err
+	}
+	streams, err = sm.stopSourceStreams(ctx, streams)
+	if err != nil {
+		return nil, err
+	}
+	positions, err := sm.syncSourceStreams(ctx, streams)
+	if err != nil {
+		return nil, err
+	}
+	return sm.verifyStreamPositions(ctx, streams, positions)
+}
+
+func (sm *streamMigrater) readSourceStreams(ctx context.Context) (map[string][]*vrStream, error) {
+	streams := make(map[string][]*vrStream)
+	var mu sync.Mutex
+	err := sm.mi.forAllSources(func(source *miSource) error {
+		stoppedStreams, err := sm.readTabletStreams(ctx, source.master, "state = 'Stopped'")
+		if err != nil {
+			return err
+		}
+		if len(stoppedStreams) != 0 {
+			return fmt.Errorf("cannot migrate until all strems are running: %s", source.si.ShardName())
+		}
+		tabletStreams, err := sm.readTabletStreams(ctx, source.master, "")
```

Comment:
This check should be done first (before checking for Stopped streams since this is a superset)

Reply:
This flow protects us from the following scenario: When we create streams, we always do it in two phases. We start them off as `Stopped`, and then update them to `Running`. If such an operation fails, we may be left with lingering `Stopped` streams. They should actually be cleaned up by the user.

In the current workflow, we stop streams and restart them. Once existing streams are stopped, there will be confusion about which of them can be restarted because they will be no different from the lingering streams.

To prevent this confusion, we first check if there are any stopped streams. If so, we request the operator to clean them up, or restart them before going ahead.

This allows us to assume that all stopped streams can be safely restarted if we cancel the operation.

I've added this as a comment to clarify.